### PR TITLE
Ledger API and sync

### PR DESCRIPTION
**Pull request by sergeuz**
### Problem

This PR supersedes https://github.com/particle-iot/device-os/pull/2660 and https://github.com/particle-iot/device-os/pull/2675.

Below are some links to simplify navigation in the new code:

- [Ledger application API](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/wiring/inc/spark_wiring_ledger.h)
- [Ledger system API](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/system/inc/system_ledger.h)
- [Ledger storage layer](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/system/src/ledger/ledger.cpp)
- [Ledger state machine](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/system/src/ledger/ledger_manager.cpp)
- [Synchronization protocol](https://github.com/particle-iot/device-os-protobuf/blob/main/cloud/ledger.proto)
- [New CoAP API](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/communication/inc/coap_api.h)
- [Variant class](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/wiring/inc/spark_wiring_variant.h)
- [Map class](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/wiring/inc/spark_wiring_map.h)

### Steps to Test

See the [validation plan](https://www.notion.so/particle/Ledger-Validation-7b64402038734c28af40b2531da8a323).

The integration tests in this PR require the device's product to have the following ledger definitions:

- `test-device-to-cloud` – type: Device to Cloud.
- `test-cloud-to-device` – type: Cloud to Device; scope: Device.

### References

- sc-118798
- sc-119316
- sc-120056
- sc-122990
- sc-124092

### CODE REVIEW

**Comment by sergeuz** [comment]
@avtolstoy @monkbroc As discussed, marking this as ready for review even though I'm still debugging the deadlock issue and need to add a few integration tests.

**Comment by avtolstoy** [comment]
I've freed up some space on Tracker and re-enabled Ledger feature flag for it.

File: communication/inc/coap_channel_new.h
Comment (monkbroc) [inline communication/inc/coap_channel_new.h:36]: Can we remove the experimental namespace? I'd rather have an explanation like "this supercedes the CoAP API xyz and should be used for new message types" than one that says "new and experimental". It won't always be new and if we're using it for real messages, it's not experimental anymore.

File: services/inc/ref_count.h
Comment (monkbroc) [inline services/inc/ref_count.h:67]: I get the semantics of this constructor and the copy constructor incrementing the reference count, and the destructor decreasing it. I don't understand the semantics of `wrap` and `unwrap` and why this class needs both interfaces.

Comment (monkbroc) [reply services/inc/ref_count.h:67]: Looking at more of the code, `wrap` and `unwrap` are used to convert between the the smart pointer and the raw pointer type, sometimes when passing the pointer to a container that deals with opaque types. That makes sense, but the usage is hard to follow because many uses of `unwrap()` are paired with a call to `release()` on the reference counted object. Ideally, every use of `wrap()` would be paired with a user of `unwrap()` and there would be no need to call `release()` directly the same way there is no need to call `addRef()` directly.

^^ The above paragraph may not be 100% accurate, but it's a note that it's hard to reason about the ownership and safety of the smart pointers above. Usually I'd look for matching pairs calls to `new/delete` or `malloc/free` in add/remove functions. Here's it's harder to do this because `wrap/unwrap` and `addRef/release` don't match.

File: communication/src/coap_channel_new.cpp
Comment (monkbroc) [inline communication/src/coap_channel_new.cpp:161]: It looks odd that you have to manually release the `RefCount` here. Is there a way to make it symmeric with `addRefToList` by using `RefCountPtr::wrap` and letting the RAII class `RefCountPtr` call `release` on the inner pointer when it goes out of scope?

File: communication/src/coap_channel_new.cpp
Comment (monkbroc) [inline communication/src/coap_channel_new.cpp:677]: What's the advantage of using a unique_ptr here since you're immediately passing ownership to the list?

File: communication/src/coap_channel_new.cpp
Comment (monkbroc) [inline communication/src/coap_channel_new.cpp:593]: This is the pattern I would expect in `removeRefFromList`. First take ownership of the reference counted object, remove it from the list, then let the smart pointer go out of scope to decrement the reference count.

File: communication/inc/coap_channel_new.h
Comment (sergeuz) [reply communication/inc/coap_channel_new.h:36]: Happy to use a different name if you have a better suggestion. `experimental` was chosen for two reasons:
- It immediately catches an eye when used in the old CoAP code where the new implementation is sometimes called in a way that breaks the separation between the existing CoAP layers. `experimental` helps to indicate that some of those changes are temporary.
- There are just too many _channel_ classes in the comms library already and isolating this one in its own namespace seemed a good way to reduce ambiguity.

File: services/inc/ref_count.h
Comment (sergeuz) [reply services/inc/ref_count.h:67]: Julien and I discussed this at length. My main takeaway from the feedback is that `release()` and `addRef()` are considered a low-level API that should not ideally be used in C++ code for which there's a smart pointer class `RefCountPtr`. I agree with that, however, I think there's at least one exception where direct usage of `release()` can be justified: if there's a reference to an object that is _not managed_ by a smart pointer in the first place, it's should be okay to call `release()` on it as it would, arguably, express the intent in a more direct and readable way as opposed to wrapping the pointer via `RefCountPtr::wrap()` and letting the resulting smart pointer immediately go out of scope.

In the code in question, `release()` is only called on objects that either came through the C-based dynalib interface or are stored in an intrusive list container where it's not possible to use smart pointers. For those objects there's always a "dangling" reference and `release()` destroys it in an explicit way.

With that said, this is quite a subjective topic. I'd be okay replacing all occurrences of `release()` with calls to `wrap()` if we find the latter more readable.

File: system/src/ledger/ledger_manager.cpp
Comment (monkbroc) [inline system/src/ledger/ledger_manager.cpp]: Is this supposed to be `contexts.end()`? This whole loop is working with a local collection of sync contexts.

File: communication/src/coap_channel_new.cpp
Comment (sergeuz) [reply communication/src/coap_channel_new.cpp:677]: I'm just being defensive here: if someone else adds another error condition before the ownership is passed to the list, they won't have a chance to leak the memory because it's already managed by a smart pointer.

File: system/src/ledger/ledger_manager.cpp
Comment (sergeuz) [reply system/src/ledger/ledger_manager.cpp]: Good catch! This code used to populate the global `contexts_` previously.

File: system/src/ledger/ledger_manager.cpp
Comment (monkbroc) [inline system/src/ledger/ledger_manager.cpp:441]: This will be important to document:
To remove the Ledger data from a device (for example, during development or to simulate the first cloud connection), go offline `Particle.disconnect()` then purge the Ledgers `Ledger::removeAll()`

Correct this if it's wrong.

Comment (sergeuz) [reply system/src/ledger/ledger_manager.cpp:441]: Just disconnecting from the cloud won't be enough if the application has valid references to the ledger instances being deleted. The most reliable way would probably be to delete ledgers early on startup, before a cloud connection is initiated and any ledger instances are created.

This is quite a limitation so I think removing it should be one of the first improvements we do the ledger API once this PR is merged.

File: system/src/ledger/ledger_manager.cpp
Comment (monkbroc) [inline system/src/ledger/ledger_manager.cpp:573]: Can you add a test in https://github.com/particle-iot/device-os-protobuf/blob/main/src/index.test.js to make sure this remains true?

Comment (sergeuz) [reply system/src/ledger/ledger_manager.cpp:573]: That's a good idea. There are integration tests in this PR that validate this indirectly but we only run them in production environment.

File: system/src/ledger/ledger_manager.cpp
Comment (monkbroc) [inline system/src/ledger/ledger_manager.cpp:822]: What would happen if a response went missing? For example, you send a subscribe request and the server sends the subscribe response but it gets lost along the way. Is the timeout mechanism at the CoAP layer? I noticed the calls to `coap_begin_request` have timeout set to 0. I'm not sure if it means "use default timeout" or "no timeout".

Comment (sergeuz) [reply system/src/ledger/ledger_manager.cpp:822]: Yeah, the old CoAP layer would take care of that and close the connection with a timeout error. 0 means the [default timeout](https://github.com/particle-iot/device-os/blob/07308ada47f7344250e7a06a281ca3d006a5f7f4/communication/inc/coap_api.h#L261) but at present it's ignored since the server only uses piggybacked CoAP responses for which only the ACK timeouts apply.

File: system/src/ledger/ledger_manager.cpp
Comment (monkbroc) [inline system/src/ledger/ledger_manager.cpp:727]: If the RTC clock was invalid here (or in other methods in this file), would the operation get retried in a subsequent run of the state machine, or would the state machine get stuck in State::SYNC_TO_CLOUD?

Comment (sergeuz) [reply system/src/ledger/ledger_manager.cpp:727]: If you take a look at the CoAP callbacks that drive most of this state machine's logic, they all transition the manager to the `FAILED` state on errors and then call into the `run()` method asynchronously that will retry the sync eventually. In this case specifically, the error from `getMillisSinceEpoch()` will be propagated to `responseCallback()` which has the above logic implemented.

File: wiring/src/spark_wiring_ledger.cpp
Comment (monkbroc) [inline wiring/src/spark_wiring_ledger.cpp:384]: On the topic of making things consistent, would we want to make all the methods to provide callbacks in the user application consistent using the pattern above? We have several patterns in the Wiring layer for callbacks, like Particle.subscribe and System.on. Could we mark some of the patterns like passing a pointer to a method and a pointer to an instance deprecated and tell developers to use lambdas? We can update the public docs as well.

Comment (sergeuz) [reply wiring/src/spark_wiring_ledger.cpp:384]: I have a minor concern related to the fact that `std::function` is not exception-safe so, strictly speaking, it should be avoided in embedded code, but otherwise I'm not opposed to the suggested change in our approach to how callbacks are managed. It would also be nice to have a discussion about the default OOM handling behavior in general as part of making that decision.

File: wiring/inc/spark_wiring_variant.h
Comment (monkbroc) [inline wiring/inc/spark_wiring_variant.h:800]: Tiny nit: can you move this to just below the definition of `VariantType` for readability?

File: wiring/src/spark_wiring_ledger.cpp
Comment (sergeuz) [reply wiring/src/spark_wiring_ledger.cpp:384]: Btw, I simplified the code that registers the `onSync` callback following your feedback in one the previous PRs. Hope it's clearer now.

File: wiring/inc/spark_wiring_map.h
Comment (monkbroc) [inline wiring/inc/spark_wiring_map.h:498]: Does `cmp_` have to be an instance variable or could it be a compile time template parameter? It seems like the map should only have to store its entries. This would bring the storage down for `VariantMap` and potentially make `Variant` more space efficient.

File: wiring/inc/spark_wiring_variant.h
Comment (monkbroc) [inline wiring/inc/spark_wiring_variant.h]: `sizeof(Variant) = 24` for now so it's worth documenting that an array of 32 bit integers will take 6x the memory when stored as a variant array.

**Review by monkbroc — APPROVED** [review]
I'm happy to mark my review as approved on this large amount of work! :clap:

File: communication/inc/coap_channel_new.h
Comment (monkbroc) [reply communication/inc/coap_channel_new.h:36]: OK. You can make the final decision here on wheter to change and to what.

File: services/inc/ref_count.h
Comment (monkbroc) [reply services/inc/ref_count.h:67]: I'm OK with keeping the usage of `release()` in those cases. I'd prefer additional comments around the ownership changes because of the bad consequences of getting it wrong for future maintainers.

File: system/src/ledger/ledger_manager.cpp
Comment (monkbroc) [reply system/src/ledger/ledger_manager.cpp:441]: It doesn't sound too bad in practice as removing a Ledger from a device isn't a normal thing that a program has to do.
Did you end up adding a Wiring API to declare which Ledgers are used by an application, so the cleanup of unused Ledgers is automatic?

File: system/src/ledger/ledger_manager.cpp
Comment (monkbroc) [reply system/src/ledger/ledger_manager.cpp:727]: Great!

File: wiring/src/spark_wiring_ledger.cpp
Comment (monkbroc) [reply wiring/src/spark_wiring_ledger.cpp:384]: Yes, I like the simplication. Thank you.

File: wiring/inc/spark_wiring_map.h
Comment (monkbroc) [reply wiring/inc/spark_wiring_map.h:498]: It's not worth doing now since the `VariantType` would stay at 24 bytes due to `String`

File: system/src/ledger/ledger_manager.cpp
Comment (sergeuz) [reply system/src/ledger/ledger_manager.cpp:441]: Yep, there's `Particle.useLedgers()` method that does that: https://github.com/particle-iot/device-os/blob/07308ada47f7344250e7a06a281ca3d006a5f7f4/wiring/inc/spark_wiring_cloud.h#L460-L472

File: communication/inc/coap_channel_new.h
Comment (sergeuz) [reply communication/inc/coap_channel_new.h:36]: I'll keep `experimental` until we have proper unit tests for this API.

File: services/inc/ref_count.h
Comment (sergeuz) [reply services/inc/ref_count.h:67]: Added more comments for now.

File: communication/src/coap_channel_new.cpp
Comment (sergeuz) [reply communication/src/coap_channel_new.cpp:161]: Discussed [here](https://github.com/particle-iot/device-os/pull/2715#discussion_r1468230265).

File: communication/src/coap_channel_new.cpp
Comment (sergeuz) [reply communication/src/coap_channel_new.cpp:593]: Discussed [here](https://github.com/particle-iot/device-os/pull/2715#discussion_r1468230265).

File: system/src/ledger/ledger_manager.cpp
Comment (sergeuz) [reply system/src/ledger/ledger_manager.cpp]: Fixed.

File: system/src/ledger/ledger_manager.cpp
Comment (sergeuz) [reply system/src/ledger/ledger_manager.cpp:573]: Added a test in device-os-protobuf: https://github.com/particle-iot/device-os-protobuf/pull/29.

File: wiring/inc/spark_wiring_variant.h
Comment (sergeuz) [reply wiring/inc/spark_wiring_variant.h:800]: Fixed.

File: wiring/inc/spark_wiring_map.h
Comment (sergeuz) [reply wiring/inc/spark_wiring_map.h:498]: > Does cmp_ have to be an instance variable or could it be a compile time template parameter?

I don't think it does as long as we don't want to support comparator objects that can be configured at runtime or are expensive to construct. I can't think of any other reason why it's stored as a member variable in STL which I used as a reference.

File: wiring/inc/spark_wiring_variant.h
Comment (sergeuz) [reply wiring/inc/spark_wiring_variant.h]: cc @rickkas7

File: wiring/inc/spark_wiring_map.h
Comment (sergeuz) [reply wiring/inc/spark_wiring_map.h:498]: Let's keep it like that for now given that we can't optimize memory usage for variants by removing it.

File: system/src/ledger/ledger.cpp
Comment (avtolstoy) [inline system/src/ledger/ledger.cpp:380]: This particular use-case of only modifying from the end should be fine for the most part though.

**Review by avtolstoy — APPROVED** [review]
Great job! There are a few future changes to Device OS that we might make to make things a bit better/simpler as discussed, hope we get to that at some point.

I did review in detail lock usage in `LedgerManager` and `Ledger`, which is probably the most likely source of problems, so far looks good, but might consider refactoring that at some point a bit just to avoid introducing problems in the future when making an accidental change.

**Comment by sergeuz** [comment]
> I did review in detail lock usage in LedgerManager and Ledger, which is probably the most likely source of problems, so far looks good, but might consider refactoring that at some point a bit just to avoid introducing problems in the future

@avtolstoy Yeah, I had to debug quite a few race conditions around that code and regretted not making `LedgerReader` and `LedgerWriter` trivial wrappers over the file handle that would simply forward all API calls to the parent ledger instance, so that we could at least simplify locking between ledger streams and instances. I'll consider that refactoring when I work on this code again.

File: hal/src/rtl872x/concurrent_hal.cpp
Comment (eberseth) [inline hal/src/rtl872x/concurrent_hal.cpp:437]: @avtolstoy  Should we submit a bug to fix in device OS on separate PR?

File: proto_defs/internal/ledger.proto
Comment (eberseth) [inline proto_defs/internal/ledger.proto:11]: Is there a need for a sequence number in the unlikely event of out-of-order Ledger sync because of slow transfer networks and modem quirks?

File: wiring/inc/spark_wiring_variant.h
Comment (eberseth) [inline wiring/inc/spark_wiring_variant.h:84]: All `float` types will have to be promoted?

File: test/unit_tests/wiring/variant.cpp
Comment (eberseth) [inline test/unit_tests/wiring/variant.cpp:207]: Check also for NaN.  I know people who use that value regularly in code to signify uninitiated values.

File: proto_defs/internal/ledger.proto
Comment (sergeuz) [reply proto_defs/internal/ledger.proto:11]: There's `update_count` that gets incremented every time the ledger is modified but it's only used internally by Device OS so that it knows which snapshot of ledger data exactly is being synchronized at a given time. It was introduced to work around a subtle race condition around [this code](https://github.com/particle-iot/device-os/blob/59325889b628f591756c40c9cc5fa33dc73c05a9/system/src/ledger/ledger.cpp#L878-L882) possible when a device-to-cloud ledger is modified by the user while it's already being synchronized with the cloud. Without the counter, the ledger manager may prematurely clear the persistent `syncPending` flag and if the device happens to reset around the same time, the last change to the ledger wouldn't be synchronized when the device restarts until the ledger is modified again.

Out-of-order syncs are not possible because only one sync can be in progress at a time. In the worst case, the system may synchronize the same ledger data twice after recovering from a disconnect that happened at the end of the previous sync attempt.

File: wiring/inc/spark_wiring_variant.h
Comment (sergeuz) [reply wiring/inc/spark_wiring_variant.h:84]: Yes. This class isn't very efficient for storing values of primitive types in the first place so it doesn't make much sense to support all possible variations of those types. 32- and 64-bit integers are handled separately because 32-bit arithmetic is done in hardware which is generally not the case for `float` and `double`.

File: test/unit_tests/wiring/variant.cpp
Comment (sergeuz) [reply test/unit_tests/wiring/variant.cpp:207]: Having a test for NaN won't hurt but I don't see how it's different from any other floating point value in this context. We're not testing serialization or anything like that here.

Comment (sergeuz) [reply test/unit_tests/wiring/variant.cpp:207]: Special floating point values are tested when serializing/deserializing a variant as CBOR, for example: https://github.com/particle-iot/device-os/blob/59325889b628f591756c40c9cc5fa33dc73c05a9/test/unit_tests/wiring/variant.cpp#L433-L441.

File: wiring/src/spark_wiring_ledger.cpp
Comment (sergeuz) [reply wiring/src/spark_wiring_ledger.cpp:384]: @avtolstoy This can be a topic to discuss with a broader group of Device OS engineers.

Comment (avtolstoy) [reply wiring/src/spark_wiring_ledger.cpp:384]: :+1: But just to say in advance this is the same pattern that we use in recently introduced APIs. GuoHui and I have gone through this during BLE introduction in detail and have come to the conclusion that function pointer with context argument + std::function wrapper most of the times is sufficient. One extra step is:
```c++
template<typename T>
void onBlah(void(T::*callback)(void arg1, void arg2), T* instance);
```

with no real dedicated implementation and just `std::bind` to produce `std::function` and delegate to its `onBlah` implementation.

**Review by avtolstoy — APPROVED** [review]
:+1:

**Comment by sergeuz** [comment]
Alright, I'm going to merge this at last. Thanks for all the feedback, everybody!

**Pull request merged**
